File: claude-code-agents-python/src/orchestrator/improvement_chain.py
"""ImprovementChain tracks quality deltas across campaign iterations."""

from __future__ import annotations

import logging
from typing import Optional

from ..models.extraction_result import ContextDelta

logger = logging.getLogger(__name__)


class ImprovementChain:
    """Accumulates ContextDelta objects across iterations and determines convergence.

    Tracks the history of improvements, computes cumulative quality changes,
    and advises whether to continue iterating based on diminishing returns
    or regression detection.

    Usage:
        chain = ImprovementChain(max_stagnant=2)
        chain.add_iteration(delta)
        if chain.should_continue():
            ...
    """

    def __init__(
        self,
        max_stagnant: int = 2,
        regression_tolerance: int = 1,
        min_improvement: float = 0.01,
    ) -> None:
        """Initialize the improvement chain.

        Args:
            max_stagnant: Max consecutive stagnant iterations before stopping.
            regression_tolerance: Max consecutive regressions before stopping.
            min_improvement: Minimum quality improvement to count as non-stagnant.
        """
        self._history: list[ContextDelta] = []
        self.max_stagnant = max_stagnant
        self.regression_tolerance = regression_tolerance
        self.min_improvement = min_improvement

    def add_iteration(self, delta: ContextDelta) -> None:
        """Add a new iteration delta to the chain.

        Args:
            delta: The context delta from the latest iteration.
        """
        self._history.append(delta)
        logger.info(
            "Iteration %d: quality %.3f -> %.3f (improvement: %+.3f)",
            delta.iteration,
            delta.quality_before,
            delta.quality_after,
            delta.quality_improvement,
        )

    def get_cumulative_delta(self) -> Optional[ContextDelta]:
        """Compute a cumulative delta summarizing all iterations.

        Returns:
            A ContextDelta summarizing the full chain, or None if empty.
        """
        if not self._history:
            return None

        all_patterns: list[str] = []
        all_failing: list[str] = []
        all_page_types: list[str] = []

        for delta in self._history:
            all_patterns.extend(delta.new_patterns)
            all_failing.extend(delta.failing_selectors)
            all_page_types.extend(delta.discovered_page_types)

        return ContextDelta(
            iteration=len(self._history),
            new_patterns=list(set(all_patterns)),
            failing_selectors=list(set(all_failing)),
            quality_before=self._history[0].quality_before,
            quality_after=self._history[-1].quality_after,
            steer_direction=self._history[-1].steer_direction,
            discovered_page_types=list(set(all_page_types)),
        )

    def should_continue(self) -> bool:
        """Determine whether the campaign should continue iterating.

        Returns False if:
        - Consecutive stagnant iterations exceed max_stagnant.
        - Consecutive regressions exceed regression_tolerance.

        Returns:
            True if the campaign should continue, False otherwise.
        """
        if not self._history:
            return True

        consecutive_stagnant = 0
        consecutive_regressions = 0

        for delta in reversed(self._history):
            improvement = delta.quality_after - delta.quality_before
            if improvement < self.min_improvement:
                consecutive_stagnant += 1
            else:
                break

        for delta in reversed(self._history):
            if delta.is_regression:
                consecutive_regressions += 1
            else:
                break

        if consecutive_stagnant >= self.max_stagnant:
            logger.info(
                "Stopping: %d consecutive stagnant iterations (max %d)",
                consecutive_stagnant,
                self.max_stagnant,
            )
            return False

        if consecutive_regressions >= self.regression_tolerance:
            logger.info(
                "Stopping: %d consecutive regressions (tolerance %d)",
                consecutive_regressions,
                self.regression_tolerance,
            )
            return False

        return True

    def get_history(self) -> list[ContextDelta]:
        """Return the full history of iteration deltas.

        Returns:
            List of ContextDelta objects in iteration order.
        """
        return list(self._history)

    @property
    def iteration_count(self) -> int:
        """Number of iterations recorded."""
        return len(self._history)

    @property
    def total_improvement(self) -> float:
        """Total quality improvement from first to last iteration."""
        if not self._history:
            return 0.0
        return self._history[-1].quality_after - self._history[0].quality_before

    @property
    def all_discovered_patterns(self) -> list[str]:
        """All unique patterns discovered across iterations."""
        patterns: set[str] = set()
        for delta in self._history:
            patterns.update(delta.new_patterns)
        return sorted(patterns)

File: claude-code-agents-typescript/src/orchestrator/improvement-chain.ts
// src/orchestrator/improvement-chain.ts — Tracks quality deltas across campaign iterations
import type { ContextDelta } from '../models/extraction-result.js';
import {
  isRegression,
} from '../models/extraction-result.js';
import { toIteration } from '../types.js';

// ── Improvement Chain ───────────────────────────────────────────
export class ImprovementChain {
  private readonly history: ContextDelta[] = [];
  private readonly maxStagnant: number;
  private readonly regressionTolerance: number;
  private readonly minImprovement: number;

  constructor(config?: {
    readonly maxStagnant?: number;
    readonly regressionTolerance?: number;
    readonly minImprovement?: number;
  }) {
    this.maxStagnant = config?.maxStagnant ?? 2;
    this.regressionTolerance = config?.regressionTolerance ?? 1;
    this.minImprovement = config?.minImprovement ?? 0.01;
  }

  addIteration(delta: ContextDelta): void {
    this.history.push(delta);
  }

  getCumulativeDelta(): ContextDelta | undefined {
    if (this.history.length === 0) return undefined;

    const allPatterns = new Set<string>();
    const allFailing = new Set<string>();
    const allPageTypes = new Set<string>();

    for (const delta of this.history) {
      for (const p of delta.newPatterns) allPatterns.add(p);
      for (const f of delta.failingSelectors) allFailing.add(f);
      for (const t of delta.discoveredPageTypes) allPageTypes.add(t);
    }

    const first = this.history[0]!;
    const last = this.history[this.history.length - 1]!;

    return {
      iteration: toIteration(this.history.length),
      newPatterns: [...allPatterns],
      failingSelectors: [...allFailing],
      qualityBefore: first.qualityBefore,
      qualityAfter: last.qualityAfter,
      steerDirection: last.steerDirection,
      discoveredPageTypes: [...allPageTypes],
    };
  }

  shouldContinue(): boolean {
    if (this.history.length === 0) return true;

    let consecutiveStagnant = 0;
    for (let i = this.history.length - 1; i >= 0; i--) {
      const delta = this.history[i]!;
      const improvement =
        (delta.qualityAfter as number) - (delta.qualityBefore as number);
      if (improvement < this.minImprovement) {
        consecutiveStagnant++;
      } else {
        break;
      }
    }

    let consecutiveRegressions = 0;
    for (let i = this.history.length - 1; i >= 0; i--) {
      const delta = this.history[i]!;
      if (isRegression(delta)) {
        consecutiveRegressions++;
      } else {
        break;
      }
    }

    if (consecutiveStagnant >= this.maxStagnant) return false;
    if (consecutiveRegressions >= this.regressionTolerance) return false;

    return true;
  }

  getHistory(): readonly ContextDelta[] {
    return [...this.history];
  }

  get iterationCount(): number {
    return this.history.length;
  }

  get totalImprovement(): number {
    if (this.history.length === 0) return 0;
    const first = this.history[0]!;
    const last = this.history[this.history.length - 1]!;
    return (last.qualityAfter as number) - (first.qualityBefore as number);
  }

  get allDiscoveredPatterns(): readonly string[] {
    const patterns = new Set<string>();
    for (const delta of this.history) {
      for (const p of delta.newPatterns) patterns.add(p);
    }
    return [...patterns].sort();
  }
}
